File: android/src/main/java/com/dodo/flutterbridge/function/FunctionNamedInvokerNode.kt
package com.dodo.flutterbridge.function

import com.dodo.flutterbridge.call.InvokerNode
import com.dodo.flutterbridge.call.strategy.InvokerStrategy
import com.dodo.flutterbridge.call.strategy.SingleInvokerStrategy
import com.dodo.flutterbridge.call.strategy.SingleInvokerStrategy.ConflictType.Exception
import com.dodo.flutterbridge.model.FlutterCallInfo
import com.dodo.flutterbridge.model.FlutterMethodInfo

/**
 *     author : liuduo
 *     e-mail : liuduo@gyenno.com
 *     time   : 2022/08/18
 *     desc   :
 *     version: 1.0
 */
class FunctionNamedInvokerNode<T> private constructor(
    override val name: String,
) : InvokerNode<T, FlutterCallInfo> {

    override val invokerStrategy: InvokerStrategy<FlutterCallInfo> =
        SingleInvokerStrategy(Exception)

    companion object {

        /**
         * 缓存同名字的FunctionNamedInvokerNode
         */
        private val namedGroups = mutableMapOf<String, FunctionNamedInvokerNode<*>>()

        /**
         * 创建或者复用一个FunctionNamedInvokerNode
         * @param name String
         * @param clazz Class<T>
         * @return FunctionNamedInvokerNode<T>
         */
        fun <T> create(name: String): FunctionNamedInvokerNode<T> {
            if (namedGroups[name] == null) {
                namedGroups[name] = FunctionNamedInvokerNode<T>(name)
            }
            return namedGroups[name] as FunctionNamedInvokerNode<T>
        }
    }

    init {
        attachInvoker(FunctionCallNode)
    }

    /**
     * encode的时候,添加名字信息
     */
    override fun encodeData(data: T): FlutterCallInfo =
        FlutterCallInfo(FlutterMethodInfo(name), data)


}

File: lib/src/function/function_named_invoker_node.dart
import 'package:flutter_bridge/src/call/invoker_node.dart';
import 'package:flutter_bridge/src/call/strategy/invoker_strategy.dart';
import 'package:flutter_bridge/src/call/strategy/single_invoker_strategy.dart';
import 'package:flutter_bridge/src/function/function_call_node.dart';
import 'package:flutter_bridge/src/model/flutter_call_info.dart';
import 'package:flutter_bridge/src/model/flutter_method_info.dart';

///   author : liuduo
///   e-mail : liuduo@gyenno.com
///   time   : 2022/08/27
///   desc   : function的name节点,处理name的逻辑
///   version: 1.0
class FunctionNamedInvokerNode<T> with InvokerNode<T, FlutterCallInfo> {

  ///缓存同名字的FunctionNamedInvokerNode
  static Map<String, FunctionNamedInvokerNode> namedGroups = {};

  ///创建或者复用一个FunctionNamedInvokerNode
  factory FunctionNamedInvokerNode.create(String name) {

    if (namedGroups[name] == null) {
      namedGroups[name] = FunctionNamedInvokerNode<T>._(name);
    }
    return namedGroups[name] as FunctionNamedInvokerNode<T>;
  }

  FunctionNamedInvokerNode._(this.name){
    attachInvoker(FunctionCallNode.instance());
  }

  @override
  InvokerStrategy<FlutterCallInfo> invokerStrategy =
      SingleInvokerStrategy(SingleInvokerConflictType.exception);

  @override
  String name;

  @override
  FlutterCallInfo encodeData(T data) =>
      FlutterCallInfo(FlutterMethodInfo(name), data);
}
